### PR TITLE
KT stabilization in PorousFlow now works with multi-procs

### PR DESCRIPTION
**Pull request by WilkAndy**
Refs #12697

### CODE REVIEW

**Comment by WilkAndy** [comment]
@dschwen , this PR is growing daily.  It'd be quite useful for me if you could review it.  You'll see it's quite repetitive and that it's based on your code.

**Comment by permcody** [comment]
> You'll see it's quite repetitive and that it's based on your code.

This is precisely why we want to make this possible in the framework. "Repetitive" and "long" is not good for maintainability. I'm ok with this going into the code in the short term, it'll help us figure out what we need to support at the framework level.

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C [hunk 1]
```diff
@@ -847,3 +841,273 @@ AdvectiveFluxCalculatorBase::PQPlusMinus(dof_id_type sequential_i,
...
+   *
+   * (B) We need to send _kij information to other processors.
+   * The same strategy is followed, first building pairs_to_receive and then
+   * _pairs_to_send.
+   */
+
+  // We use this (hopefully unique) message tag for all communication in this method.
+  // QUESTION: If there are multiple AdvectiveFluxCalculatorBase in the input file, will this
+  // work?
+  Parallel::MessageTag send_tag = _communicator.get_unique_tag(4712);
```
Comment (dschwen) [inline modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C]: Did you just increment my choice of tag here? It was based on
image

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C [hunk 2]
```diff
@@ -847,3 +841,273 @@ AdvectiveFluxCalculatorBase::PQPlusMinus(dof_id_type sequential_i,
...
+  std::vector<Parallel::Request> send_requests(_app.n_processors() - 1);
+  unsigned sr = 0;
+  for (processor_id_type pid = 0; pid < _app.n_processors(); ++pid)
+    if (pid != _my_pid)
+      // non-blocking send
+      _communicator.send(pid, _nodes_to_receive[pid], send_requests[sr++], send_tag);
+
+  // Now receive the adverts from all other processors
+  // QUESTION: what is item_type?  Is there a better way of setting it?
+  const auto item_type = libMesh::Parallel::StandardType<dof_id_type>(&(_nodes_to_receive[0][0]));
```
Comment (dschwen) [inline modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C]: As far as I understand item type tells MPI about the type layout of the data you are trying to communicate. In a heterogeneous environment you could probably have MPI transfer everything as byte buffers. iten_type will make MPI aware of the actual types and will make it take care of little/big endian conversions for example.

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C]: Yea, i just incremented your choice.  I was trying to think "4" was "D" (for Daniel) but then couldn't get the "7", "11, or "71", etc match anything sensible.

**Comment by WilkAndy** [comment]
Please could someone review this, ready for a merge.

**Comment by WilkAndy** [comment]
Unrelated heavy-test failure: xfem timeout

**Comment by friedmud** [comment]
This isn't the right way to do this at all, and the framework team has a plan to fix this that won't require any communication.  However, no one is available to implement that at the moment... so a compromise is needed.

I'm willing to allow communication here... but it needs to use much better semantics for the neighbor exchange.

The functions you should be using are: `push_parallel_vector_data()` and `pull_parallel_vector_data()` which you can find here:

https://github.com/libMesh/libmesh/blob/master/include/parallel/parallel_sync.h#L62

Those are scalable neighbor exchange algorithms that @roystgnr and myself have worked hard on implementing.

Using these there should be a HUGE reduction in the amount of code you have here.

If you have any questions, let me know - the comments on those function declarations are pretty explanatory though...

**Comment by WilkAndy** [comment]
Thanks @friedmud for your mention of `push_parallel_vector_data()`, etc.

**Comment by WilkAndy** [comment]
I used @friedmud 's `push_parallel_vector_data` and it reduced the size of this PR nicely, and it's more readable too.  Thanks @friedmud .

**Review by permcody — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for the help with this, everyone.   A very long-standing issue finally solved !!

**Pull request merged**